crypto/evp/evp_rand.c

```diff
@@ -440,7 +439,8 @@ int EVP_RAND_generate(EVP_RAND_CTX *ctx, unsigned char *out, size_t outlen,
         params[0] = OSSL_PARAM_construct_size_t(OSSL_DRBG_PARAM_MAX_REQUEST,
                                                 &chunk);
         params[1] = OSSL_PARAM_construct_end();
-        if (!EVP_RAND_get_ctx_params(ctx, params) || chunk == 0) {
+        /* Cannot call EVP_RAND_get_ctx_params() since we have the lock */
+        if (!ctx->meth->get_ctx_params(ctx->data, params) || chunk == 0) {
```

Comment:
err yuk :)

Reply:
Why?  It's just calling a method.

Reply:
I dont think you have any other options here :)

Reply:
> I dont think you have any other options here :)

There is _always_ another option:) This change here fixes a single recursive lock at the prise of readability. The next recursive lock might be lurking elsewhere. IMHO there is a more generic and consistent solution to avoid taking recursive locks: identify _all_ EVP_RAND functions which take the lock and split them into a lowercase private (inline?) function which does the actual work

- `evp_rand_generate()`, `evp_rand_get_ctx_params()`, ...

and a public uppercase function which wraps the locking around the lowercase function.

- `EVP_RAND_generate()`, `EVP_RAND_get_ctx_params()`, ...

Then stick to rule to always call the lowercase functions internally. That' s how I did it in RAND_DRBG. It comes at the prise of an extra function call which in many cases the compiler can inline away.

Reply:
> identify all EVP_RAND functions

ok, not really all, but all funcitons which need to be called internally. It's not worth to split `EVP_RAND_generate()` if it does not get called internally anyway.

Reply:
If this solution is adopted, it should apply to *all* EVP_RAND functions, this should include `EVP_RAND_generate()`.

BTW: the mixed case functions tripped me up a couple of times.

This sounds like it should be an OTC coding style decision.

Reply:
All refactored.  I used lower case function names with a __locked_ suffix to better indicate what the functions do.